proposals/2117-marquez-over-time.md

```diff
@@ -0,0 +1,29 @@
+# Proposal: Marquez API to retrieve historical data
+
+## Problem
+
+Marquez data model in PostgresSQL allows extracting a snapshot of any lineage content from the past. This may be extremely
```

Comment:
Is the intention here to return _lineage_ for a given point in time? Or just the dataset/job definitions as they were at a particular version? If the latter, don't we already have that? If the former, do you intend to store the lineage information differently? The current storage model is sufficient to return point-in-time lineage, but it's really slow if looking at anything but the latest version.

Reply:
The intention of this PR is to make consensus on how do we expose point in time endpoints, including lineage endpoints as well. Storage model, and performance related it, is specific to each endpoint and it's outside the scope of this PR. You're right that it may require some remodeling but I think this should be discussed one by one when implementing specific endpoint for point in time. 

In other words: this PR shows how API for lineage over time should look like, but not how to implement it.

Reply:
I'm fine with agreeing on the API changes before tackling the storage, but I do think there ought to be a design doc tackling the storage issue up front. We're really only talking about the `/lineage` and `/column-lineage` APIs here, so it's really a question of how to store point-in-time lineage in a scalable way.

Reply:
I agree with @collado-mike, to get a snapshot of a lineage graph the caller _should_ invoke `GET` `/lineage` with a `nodeId` (see proposed `nodeId`s below). A lineage graph is modified within the context of a run. We can call these incremental changes to the graph at a given point in time via a some run `R` as a  **run-level** graph (=lineage graph snapshot). A run-level graph represents the relationships between dataset, job, and run metadata at a given point in time.

### Graph Data Model

**A run-level graph consists of the following nodes:**

* **Dataset Version:** A read-only _immutable_ version of a dataset.
* **Job Version:** A read-only _immutable_ version of a job, with a unique referenceable link to code preserving the reproducibility of builds from source.
* **Run:** A discrete instantiation of a job version, with a unique run ID used to update each stage of execution.

### Nodes

| **ID**       | **`dataset:{namespace}:{dataset}#{version}`**                     |
|:-------------|:------------------------------------------------------------------|
| **Example**  | **`dataset:food_delivery:public.top_delivery_times#947c0388..`**  |

| **ID**       | **`job:{namespace}:{job}#{version}`**                             |
|:-------------|:------------------------------------------------------------------|
| **Example**  | **`job:food_delivery:orders_popular_day_of_week#947c0388..`**     |

| **ID**       | **`run:{id}`**                                                    |
|:-------------|:------------------------------------------------------------------|
| **Example**  | **`run:a03422cf..`**